### ISSUE TITLE
mingw-w64 builds fail due to "export ordinal too large"

### ISSUE THREAD

**Issue by spl1inter** [IfcOpenShell/IfcOpenShell#1725]
Hi there,

It's been a little while and I though I would come check how this is going.

Surprisingly everything builds out of the box, but the linking is where it seems to get into trouble with some versions of MinGW, using various tools etc. Here are a few things probably at least worth mentioning:

Linking a monolithic libIfcParse.dll dies with "export ordinal too large. Basically it means that the dll is trying to export more than 65000 symbols in a single table I think. Looking at the debating packages I see that there was an attempt at a non-monolithic which subdivided 2x3 from 4 but mutual dependencies between the various ifcGeom dlls made it unbuildable.

Also it might be worth mentioning about how you do your symbol exports, for example the _dllexport attribute on template instances on-the-fly in some releases, sorry I don't have that file anymore but essentially the issue was with clang++ ignoring the export attribute when you do something like this:
IFC_GEOM_API class ...<..>
And symbols were missing anyway from the dlls when I finally did massage it to build. Again, for some reason in these cases I have better luck when packages use -Wl,--export-all-symbols *.obj as opposed to -Wl,--whole-archive *.a

I don't know if the feedback is useful in any way but I figured it would be better to give it than not.

**Comment by aothms** [comment]
> Linking a monolithic libIfcParse.dll dies with "export ordinal too large".. Basically it means that the dll is trying to export more than 65000 symbols in a single table I think. Looking at the debian packages I see that there was an attempt at a non-monolithic which subdivided 2x3 from 4 but mutual dependencies between the various ifcGeom dlls made it unbuildable.

Yes, this is also the case in https://github.com/IfcOpenShell/IfcOpenShell/issues/1029 The same applies when building a DLL on MSVC (but not a static lib). For now the workaround is to reduce the amount of schemas in IfcParse https://github.com/IfcOpenShell/IfcOpenShell/issues/1029#issuecomment-871650453 You likely don't need them all anyway.

> the _dllexport attribute on template instances on-the-fly

In the 0.7 branch I have went ahead and removed most of the templates in the geom namespace actually. They weren't very useful anyway.

> Anyway these are very specific linking issues limited to very niche tools. I don't know if the feedback is useful in any way but I figured it would be better to give it than not.

It's useful, but as you say the underlying issue is probably this monolithic IfcParse library. A plug-in architecture is probably better where individual schemas are compiled into shared libs and are only imported at runtime when a file with a schema by that name is encountered.

I think IfcOpenShell is at a point of maturity where these kind of issues exactly due matter. And while there is a wealth of contributions on the python side. On C++ (and especially these more fundamental issues) we can use a lot of help.

**Comment by spl1inter** [comment]
Thanks that did get me further, provided I am content to only have two schemas enabled :) I wasn't sure I would run into the same missing interdependencies as I mentioned above with that but you are right, 0.7.0 definitely addressed a lot of the  dllexport-ugliness in 0.6.0 if the builder is willing to work through all the packaging issues..

1) multiple fixes / adaptations to CMakeList.txt so that szip.dll.a is correctly discovered, aec.a isn't expected and nothing looks for an MSVC style *.lib* import file... 

2) Manual editing of all generated link.txt to hold pattern similar to:
`g++.exe -O3 -DNDEBUG -O3 -shared -o libIfcParse.dll -Wl,--out-implib,libIfcParse.dll.a -Wl,--major-image-version,0,--minor-image-version,6 -Wl,--export-all-symbols @CMakeFiles/ifcParse.dir/objects1.rsp -Wl,--no-whole-archive @CMakeFiles\IfcParse.dir\linklibs.rsp`
This is due to certain needed symbols still seem to be missing from the export table (IfcParse::schema_definition::instantiate is a common culprit for MinGW, even on previous versions of IFCOS..)
Could addressed in source but I was in a hurry..

3) issues with symbol conflicts between multiple libIfcGeom_ifcxxx.a when multiple api are enabled. Right now I link libIfcGeom.dll with -Wl,--allow-multiple-definition, which almost certainly wrong. Why the same namespace is being exported from more than one archive totally escapes me..

4) For things like Serializer and Convert, this link order is wrong for MinGW (maybe others):
 libSerializers_ifc2x3.a libSerializers_ifc4x3_rc4.a libIfcGeom.dll.a libIfcGeom_ifc2x3.a libIfcGeom_ifc4x3_rc4.a libIfcParse.dll.a
When static libs are involved under MinGW it _must_ be:
 libSerializers_ifc2x3.a libSerializers_ifc4x3_rc4.a libIfcGeom_ifc2x3.a libIfcGeom_ifc4x3_rc4.a libIfcGeom.dll.a libIfcParse.dll.a
(once libIfcGeom.dll.a libIfcParse.dll.a themselves have been built correctly as in the previous step.. without which its _still_ won't work.. so #2 is really key..)

5) suggest installing include-lib/serializers as include-lib/ifcSerializers a) to match the rest of the package and b) prevent conflict with other generically named system includes or manually need to uninstall (cmake doesn't offer a reliable uninstall target). I don't see the advantage of the current naming of this..

That's my feedback then I hope it helps..

**Comment by aothms** [comment]
> multiple fixes / adaptations to CMakeList.txt so that szip.dll.a

Yes the HDF5 bit is still rather last minute and sloppy. If you see a way to make some quick fixes (I can imagine NOT UNIX is equated with MSVC for example) that'd be greatly appreciated :)

> IfcParse::schema_definition::instantiate is a common culprit for MinGW

I've added export on `IfcParse::schema_definition`, apparently was missing. b9084ad0f5069dc6a66a57c0be2ca56e909d2c22

> Why the same namespace is being exported from more than one archive totally escapes me..

Because in 0.5 the whole of IfcOpenShell would only target 1 IFC schema. In 0.6 we had to separate the schema agnostic parts from the schema specific parts. We just stopped when it was working for the systems we regularly test on. Can you name a couple of symbols for which this is the case?

> For things like Serializer and Convert, this link order is wrong for MinGW (maybe others):

It works for clang and gcc, so a bit odd that it doesn't work for the mingw linker. Actually we're pretty lazy in this regard we just repeat until it works e.g https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/cmake/CMakeLists.txt#L572 `IfcGeom ${IFCGEOM_SCHEMA_LIBRARIES} IfcGeom ${IFCGEOM_SCHEMA_LIBRARIES}` and https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/cmake/CMakeLists.txt#L308 `${OPENCASCADE_LIBRARIES} ${OPENCASCADE_LIBRARIES} ${OPENCASCADE_LIBRARIES} ${OPENCASCADE_LIBRARIES} ${OPENCASCADE_LIBRARIES}`. Yes 5 repititions before all cyclic dependencies finally got resolved in all cases. So maybe we can solve this in this way too, would be great if you can test and do a quick edit.

> suggest installing include-lib/serializers as include-lib/ifcSerializers

That's a good comment. But I'll probably solve it in another way for 0.7, we have quite a bit of renaming we want to do to eliminate confusion between classes from the IFC schema and classes from IfcOpenShell. It will probably become something like `ifcopenshell/serializers` and reserve the Ifc prefix solely for constructs from the IFC schema.

**Comment by spl1inter** [comment]
> Can you name a couple of symbols for which this is the case?

I can do that. Wl,--allow-multiple-definition temporarily removed from ifcGeom.dir/linklibs.rsp with system prefixes removed:

```
<PREFIX>mingw32/bin/ld.exe: libIfcGeom_ifc4x3_rc4.a(IfcGeomRepresentation.cpp.obj):IfcGe
omRepresentation.cpp:(.text+0x310): multiple definition of `IfcGeom::Representat
ion::BRep::calculate_surface_area(double&) const'; libIfcGeom_ifc2x3.a(IfcGeomRe
presentation.cpp.obj):IfcGeomRepresentation.cpp:(.text+0x310): first defined here
<PREFIX>mingw32/bin/ld.exe: libIfcGeom_ifc4x3_rc4.a(IfcGeomRepresentation.cpp.obj):IfcGe
omRepresentation.cpp:(.text+0x400): multiple definition of `IfcGeom::Representat
ion::BRep::calculate_volume(double&) const'; libIfcGeom_ifc2x3.a(IfcGeomRepresen
tation.cpp.obj):IfcGeomRepresentation.cpp:(.text+0x400): first defined here
<PREFIX>mingw32/bin/ld.exe: libIfcGeom_ifc4x3_rc4.a(IfcGeomRepresentation.cpp.obj):IfcGe
omRepresentation.cpp:(.text+0x510): multiple definition of `apply_transformation
(TopoDS_Shape const&, gp_Trsf const&)'; libIfcGeom_ifc2x3.a(IfcGeomRepresentatio
n.cpp.obj):IfcGeomRepresentation.cpp:(.text+0x510): first defined here
<PREFIX>mingw32/bin/ld.exe: libIfcGeom_ifc4x3_rc4.a(IfcGeomRepresentation.cpp.obj):IfcGe
omRepresentation.cpp:(.text+0x7c0): multiple definition of `apply_transformation
(TopoDS_Shape const&, gp_GTrsf const&)'; libIfcGeom_ifc2x3.a(IfcGeomRepresentati
on.cpp.obj):IfcGeomRepresentation.cpp:(.text+0x7c0): first defined here
<PREFIX>mingw32/bin/ld.exe: libIfcGeom_ifc4x3_rc4.a(IfcGeomRepresentation.cpp.obj):IfcGe
omRepresentation.cpp:(.text+0xa30): multiple definition of `IfcGeom::Representat
ion::BRep::as_compound(bool) const'; libIfcGeom_ifc2x3.a(IfcGeomRepresentation.c
pp.obj):IfcGeomRepresentation.cpp:(.text+0xa30): first defined here
<PREFIX>mingw32/bin/ld.exe: libIfcGeom_ifc4x3_rc4.a(IfcGeomRepresentation.cpp.obj):IfcGe
omRepresentation.cpp:(.text+0xdd0): multiple definition of `IfcGeom::Representat
ion::Serialization::Serialization(IfcGeom::Representation::BRep const&)'; libIfc
Geom_ifc2x3.a(IfcGeomRepresentation.cpp.obj):IfcGeomRepresentation.cpp:(.text+0x
dd0): first defined here
<PREFIX>mingw32/bin/ld.exe: libIfcGeom_ifc4x3_rc4.a(IfcGeomRepresentation.cpp.obj):IfcGe
omRepresentation.cpp:(.text+0xdd0): multiple definition of `IfcGeom::Representat
ion::Serialization::Serialization(IfcGeom::Representation::BRep const&)'; libIfc
Geom_ifc2x3.a(IfcGeomRepresentation.cpp.obj):IfcGeomRepresentation.cpp:(.text+0x
dd0): first defined here
<PREFIX>mingw32/bin/ld.exe: libIfcGeom_ifc4x3_rc4.a(IfcGeomRepresentation.cpp.obj):IfcGe
omRepresentation.cpp:(.text+0x2560): multiple definition of `IfcGeom::Representa
tion::BRep::calculate_projected_surface_area(gp_Ax3 const&, double&, double&, do
uble&) const'; libIfcGeom_ifc2x3.a(IfcGeomRepresentation.cpp.obj):IfcGeomReprese
ntation.cpp:(.text+0x2560): first defined here
collect2.exe: error: ld returned 1 exit status
```

Unfortunately I can't test the final binaries because the python API has changed slightly since FreeCAD 18 and I don't know enough about the history of thise changes. But the tools all seem to work..

**Comment by BIMvoice** [comment]
Housekeeping: this is a 2021 MinGW-w64 linking report (duplicate per-schema symbols / export ordinal overflow) against the 0.6-era build layout. The C++ side has been substantially restructured since, the referenced link flags and library split no longer match the current tree, and there has been no activity in five years. Recommend closing as obsolete; a fresh report against a current build would be the way to resurrect MinGW support.
